src/graph/executor/algo/AllPathsExecutor.cpp

```diff
@@ -0,0 +1,461 @@
+  for (auto& vid : leftInitVids_) {
+    auto vidIter = leftAdjList_.find(vid);
+    if (vidIter == leftAdjList_.end()) {
+      continue;
+    }
+    auto src = vidIter->first;
+    auto& adjEdges = vidIter->second;
+    if (adjEdges.empty()) {
+      continue;
+    }
+    pathsPtr->reserve(adjEdges.size() + pathsPtr->size());
+    for (auto& edge : adjEdges) {
+      pathsPtr->emplace_back(std::vector<Value>({src, edge}));
+    }
+  }
+  size_t step = 2;
+  auto future = doBuildPath(step, 0, pathsPtr->size(), pathsPtr);
+  return future.via(runner()).thenValue([this](std::vector<Row>&& paths) {
+    memory::MemoryCheckGuard guard;
+    if (!paths.empty()) {
+      result_.rows.swap(paths);
+    }
+    return Status::OK();
+  });
+}
+
+folly::Future<std::vector<Row>> AllPathsExecutor::doBuildPath(
+    size_t step,
+    size_t start,
+    size_t end,
+    std::shared_ptr<std::vector<std::vector<Value>>> pathsPtr) {
+  if (cnt_.load(std::memory_order_relaxed) >= limit_) {
+    return folly::makeFuture<std::vector<Row>>(std::vector<Row>());
+  }
+
+  auto& adjList = leftAdjList_;
+  auto currentPathPtr = std::make_unique<std::vector<Row>>();
+  auto newPathsPtr = std::make_shared<std::vector<std::vector<Value>>>();
+
+  for (auto i = start; i < end; ++i) {
+    auto& path = (*pathsPtr)[i];
+    auto& edgeValue = path.back();
+    DCHECK(edgeValue.isEdge());
+    auto& dst = edgeValue.getEdge().dst;
+    auto dstIter = rightInitVids_.find(dst);
+    if (dstIter != rightInitVids_.end()) {
+      Row row;
+      row.values.emplace_back(path.front());
+      List edgeList(std::vector<Value>(path.begin() + 1, path.end()));
+      row.values.emplace_back(std::move(edgeList));
+      row.values.emplace_back(*dstIter);
+      currentPathPtr->emplace_back(std::move(row));
+      ++cnt_;
+      if (cnt_.load(std::memory_order_relaxed) >= limit_) {
+        break;
+      }
+    }
+    if (step <= maxStep_) {
+      auto adjIter = adjList.find(dst);
+      if (adjIter == adjList.end()) {
+        continue;
+      }
+
+      auto& adjedges = adjIter->second;
+      for (auto& edge : adjedges) {
+        if (noLoop_) {
+          if (hasSameVertices(path, edge.getEdge())) {
+            continue;
+          }
+        } else {
+          if (hasSameEdge(path, edge.getEdge())) {
+            continue;
+          }
+        }
+        // copy
+        auto newPath = path;
+        newPath.emplace_back(adjIter->first);
+        newPath.emplace_back(edge);
+        newPathsPtr->emplace_back(std::move(newPath));
+      }
+    }
+  }
+
+  auto newPathsSize = newPathsPtr->size();
+  if (step > maxStep_ || newPathsSize == 0) {
+    return folly::makeFuture<std::vector<Row>>(std::move(*currentPathPtr));
+  }
+  std::vector<folly::Future<std::vector<Row>>> futures;
+  if (newPathsSize < FLAGS_path_batch_size) {
+    futures.emplace_back(folly::via(runner(), [this, step, newPathsSize, newPathsPtr]() {
+      return doBuildPath(step + 1, 0, newPathsSize, newPathsPtr);
```

Comment:
seems this may use many memory, try to throw bad_alloc here, make sure it failed with expect GraphMemoryExceed error.

Reply:
ACK